downstairs/src/region.rs

```diff
@@ -758,27 +758,40 @@ impl Region {
         Ok(result)
     }
 
-    pub fn validate_hashes(
+    /// Checks that the hashes are valid for all of the input writes
+    ///
+    /// # Panics
+    /// If any write is structurally invalid, i.e. having a different number of
+    /// blocks and block contexts.
+    fn validate_hashes(
         &self,
-        writes: &[crucible_protocol::Write],
+        write: &RegionWrite,
     ) -> Result<(), CrucibleError> {
-        for write in writes {
-            let computed_hash = if let Some(encryption_context) =
-                &write.block_context.encryption_context
+        let block_size = self.def().block_size() as usize;
+        for req in write.iter() {
+            // TODO do some of `check_input` here instead of panicking?
+            let w = &req.write;
+            if w.data.len() / block_size != w.block_contexts.len() {
+                panic!("invalid write; block count must match context count");
+            }
+            for (block, ctx) in w.data.chunks(block_size).zip(&w.block_contexts)
             {
-                integrity_hash(&[
-                    &encryption_context.nonce[..],
-                    &encryption_context.tag[..],
-                    &write.data[..],
-                ])
-            } else {
-                integrity_hash(&[&write.data[..]])
-            };
+                let computed_hash =
+                    if let Some(encryption_context) = &ctx.encryption_context {
+                        integrity_hash(&[
+                            &encryption_context.nonce[..],
+                            &encryption_context.tag[..],
+                            &block,
+                        ])
+                    } else {
+                        integrity_hash(&[&block])
+                    };
 
-            if computed_hash != write.block_context.hash {
-                error!(self.log, "Failed write hash validation");
-                // TODO: print out the extent and block where this failed!!
-                crucible_bail!(HashMismatch);
+                if computed_hash != ctx.hash {
+                    error!(self.log, "Failed write hash validation");
+                    // TODO: print out the extent and block where this failed!!
```

Comment:
[I opened this issue ](https://github.com/oxidecomputer/crucible/issues/1307) for this.  In the event that we do get a hash validation error, all the data we can dump can only help us figure out what has happened.